### PR TITLE
TemplateFactory: add support for custom Template implementation

### PR DESCRIPTION
**Pull request by mrtnzlml**
With this change it's possible to simply use custom Template implementation. Before this commit Template object initialization was hard-coded into TemplateFactory (related issue: https://github.com/nette/application/issues/141). Usage example (`config.neon`):

```
latte:
    templateClass: App\CustomTemplateImplementation
```

Custom template implementation MUST implement Nette\Application\UI\ITemplate interface.
- feature
- issues - #141
- documentation - probably not needed because this part of framework is not documented (internals)
- BC break - no

This is more request for comments than real PR. Thank you.

### CODE REVIEW

File: src/Bridges/ApplicationLatte/TemplateFactory.php [hunk 1]
```diff
@@ -31,14 +31,20 @@ class TemplateFactory implements UI\ITemplateFactory
...
 	public function __construct(ILatteFactory $latteFactory, Nette\Http\IRequest $httpRequest = NULL,
-		Nette\Security\User $user = NULL, Nette\Caching\IStorage $cacheStorage = NULL)
+		Nette\Security\User $user = NULL, Nette\Caching\IStorage $cacheStorage = NULL, $templateClassName = Template::class)
 	{
 		$this->latteFactory = $latteFactory;
 		$this->httpRequest = $httpRequest;
 		$this->user = $user;
 		$this->cacheStorage = $cacheStorage;
+
+		\Nette\Utils\Validators::assert($templateClassName, 'string');
```
Comment (TomasVotruba) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: This is not needed, as it is not convention in Nette.
Also PHP 7 is comming soon with type checks.

File: src/Bridges/ApplicationLatte/TemplateFactory.php [hunk 2]
```diff
@@ -48,7 +54,11 @@ public function __construct(ILatteFactory $latteFactory, Nette\Http\IRequest $ht
 	public function createTemplate(UI\Control $control = NULL)
 	{
 		$latte = $this->latteFactory->create();
-		$template = new Template($latte);
+		$template = new $this->templateClassName($latte);
```
Comment (TomasVotruba) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: This looks ok. Just wondering about use case, when custom class has different arguments.

File: tests/Bridges.Latte/TemplateFactory.customTemplate.phpt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+
+$factory = new TemplateFactory(new LatteFactoryMock(new Latte\Engine), NULL, NULL, NULL, TemplateMock::class);
+$template = $factory->createTemplate();
+Assert::type(TemplateMock::class, $template);
+Assert::type(UI\ITemplate::class, $template);
+Assert::same([], $template->flashes);
+Assert::same('ok', $template->render());
+$template->setFile('bla');
+Assert::same('alb', $template->render());
```
Comment (TomasVotruba) [inline tests/Bridges.Latte/TemplateFactory.customTemplate.phpt]: Awesome test, thank you.

File: src/Bridges/ApplicationLatte/TemplateFactory.php (hunk 1 above)
Comment (mrtnzlml) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: It's not convention in Nette doesn't mean I cannot do that, does it? And it's not true that it's not needed (before PHP 7). Moreover PHP 7 is not used in Nette yet.

File: src/Bridges/ApplicationLatte/TemplateFactory.php (hunk 2 above)
Comment (mrtnzlml) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: It should be in another PR if needed I think. It's not possible to add another arguments with current implementation as well.

File: tests/Bridges.Latte/TemplateFactory.customTemplate.phpt [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	}
+}
+
+class TemplateMockWithoutImplement {}
+
+class TemplateMock implements UI\ITemplate
+{
+	private $file = 'ko';
+
+	public function render() {
```
Comment (mrtnzlml) [inline tests/Bridges.Latte/TemplateFactory.customTemplate.phpt]: TODO: fix coding standards

File: src/Bridges/ApplicationLatte/TemplateFactory.php (hunk 1 above)
Comment (TomasVotruba) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: Code would be inconsistent with other strings (scalar) types in methods. 

Rather check if this class implements `ITemplate` interface. Because it would fail later and now it silently pass in here creating invalid object.

File: src/Bridges/ApplicationLatte/TemplateFactory.php (hunk 2 above)
Comment (TomasVotruba) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: Ok, let's skip it now then.

Btw, do you need `$latte` as argument in your implementation?

File: src/Bridges/ApplicationLatte/TemplateFactory.php (hunk 1 above)
Comment (mrtnzlml) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: I have to do both checks because without checking string it's possible to call `new` on integer for example. From my point of view validation is needed. Can you help me how to prevent calling `$template = new $this->templateClassName($latte);` (line 57) with wrong `$this->templateClassName` type?

Comment (TomasVotruba) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: I see. If so, add them both here. It doesn't make sense splitting them.
Standalone method would be best, e.g. `validateTemplateClassName()`.

File: src/Bridges/ApplicationLatte/TemplateFactory.php (hunk 2 above)
Comment (mrtnzlml) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: Not necessarily because I am extending `Nette\Bridges\ApplicationLatte\Template` and then calling `$this->getLatte()`. So the answer is no but original implementation needs it.

File: src/Bridges/ApplicationLatte/TemplateFactory.php (hunk 1 above)
Comment (mrtnzlml) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: Ok, I'll improve it in next iteration. I am used to doing validations in constructor in my projects...

Comment (TomasVotruba) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: Yes. This should be in constructor.

Something like this:

```php
public function __construct($className = Tempalte::class)
{
    $this->validateTempalteClassName($className);
}
```

Comment (mrtnzlml) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: I don't see benefit here. It's only because of calling line `Nette\Utils\Validators::assert($templateClassName, 'string');` here. I don't like it... :)

Waiting for more opinions...

Comment (TomasVotruba) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: We missunderstood a bit. I've updated the code.

Comment (mrtnzlml) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: Yeah, it cannot be done like that. It's because `$template` is created during `createTemplate` call (everytime you call it) and therefore you cannot test its instance in constructor. You don't have `ITemplate` in constructor. What implements `ITemplate`? The answer is: created `$template`.

**Comment by JanTvrdik** [comment]
Setter (as suggested in #141) is IMHO better (template class name has a character of *configuration* not of *dependency*)

**Comment by mrtnzlml** [comment]
@JanTvrdik I love [this idea](https://github.com/nette/application/issues/141#issuecomment-249413992) (despite the naming - just idea) even more.

**Comment by JanTvrdik** [comment]
You mean the „TemplateClassFactory“ which would create just instance of `Template` without configuring it any further? That seems like a terrible idea to me.

Good solutions are either setter or splitting the instance creation and configuration to multiple methods to allow overriding part of the logic.

**Comment by TomasVotruba** [comment]
**I agree with @JanTvrdik in setter as configuration.**

If you want to change class name, you might want to do it multiple times. With constructor dependency you would have to create new instance of TemplateFactorry.

**Comment by mrtnzlml** [comment]
TODO for me: Change implementation back to the setter one. Make it configurable using DI extension if possible (because original usage is not nice).

**Comment by mrtnzlml** [comment]
Ok guys, what about this implementation?

File: src/Bridges/ApplicationDI/LatteExtension.php [hunk 5]
```diff
@@ -19,6 +19,7 @@ class LatteExtension extends Nette\DI\CompilerExtension
 	public $defaults = [
 		'xhtml' => FALSE,
 		'macros' => [],
+		'templateClass' => \Nette\Bridges\ApplicationLatte\Template::class
```
Comment (mrtnzlml) [inline src/Bridges/ApplicationDI/LatteExtension.php]: Is this coupling between bridges fine?

File: src/Bridges/ApplicationLatte/TemplateFactory.php [hunk 6]
```diff
@@ -48,7 +51,11 @@ public function __construct(ILatteFactory $latteFactory, Nette\Http\IRequest $ht
 	public function createTemplate(UI\Control $control = NULL)
 	{
 		$latte = $this->latteFactory->create();
-		$template = new Template($latte);
+		$template = new $this->templateClass($latte);
+		if (!$template instanceof UI\ITemplate) {
```
Comment (JanTvrdik) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: This check should be moved to `setTemplateClass` (see is_a with third argument set to TRUE)

File: src/Bridges/ApplicationLatte/TemplateFactory.php [hunk 7]
```diff
@@ -112,4 +119,10 @@ public function createTemplate(UI\Control $control = NULL)
 		return $template;
 	}
 
+	public function changeTemplateClass($templateClass)
```
Comment (JanTvrdik) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: should be named setTemplateClass (due to consistency)

File: src/Bridges/ApplicationLatte/TemplateFactory.php [hunk 8]
```diff
@@ -112,4 +119,10 @@ public function createTemplate(UI\Control $control = NULL)
...
+	{
+		\Nette\Utils\Validators::assert($templateClass, 'string');
```
Comment (JanTvrdik) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: should be removed (due to consistency)

File: src/Bridges/ApplicationDI/LatteExtension.php (hunk 5 above)
Comment (JanTvrdik) [inline src/Bridges/ApplicationDI/LatteExtension.php]: * should be simplified to `Nette\Bridges\ApplicationLatte\Template::class`
* the usecase se imho too narrow to add option to the extension so I would remove this entirely
* if added, the default value should be `NULL` and the setter call should only be emitted if the value is not NULL

File: src/Bridges/ApplicationLatte/TemplateFactory.php (hunk 6 above)
Comment (JanTvrdik) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: You will also probably need to call `class_exists` on the `$templateClass` to load the class to memory

File: src/Bridges/ApplicationLatte/TemplateFactory.php (hunk 8 above)
Comment (mrtnzlml) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: How would you prevent passing something different than string? I understand your note about consistency, but it's wrong I think. Or should I remove usage of `Validators` and use custom check?

File: src/Bridges/ApplicationLatte/TemplateFactory.php (hunk 6 above)
Comment (mrtnzlml) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: Thanks - I didn't know the third argument of is_a.

**Comment by mrtnzlml** [comment]
Updated - I have no idea why is old conversation still here (what a mess in this discussion)... :-/
What changed:

- it uses setter implementation
- it's still configurable via Latte DI extension
- it's still used validation in `setTemplateClass` because I strongly believe that we should check if passed class name is string (and not created instance for example)

Thanks for help.

File: src/Bridges/ApplicationLatte/TemplateFactory.php (hunk 8 above)
Comment (JanTvrdik) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: > How would you prevent passing something different than string? 

Here is the secret – you don't prevent passing something different because this is PHP because it would kill code readability and performance. Pretty much every Nette method without typehint can be broken by passing the wrong type.

File: src/Bridges/ApplicationLatte/TemplateFactory.php [hunk 9]
```diff
@@ -112,4 +115,13 @@ public function createTemplate(UI\Control $control = NULL)
 		return $template;
 	}
 
+	public function setTemplateClass($templateClass)
+	{
+		\Nette\Utils\Validators::assert($templateClass, 'string');
+		if (!is_a($templateClass, UI\ITemplate::class, TRUE)) {
+			throw new \Nette\Utils\AssertionException('Template should be instance of ' . UI\ITemplate::class . ", $templateClass given.");
```
Comment (JanTvrdik) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: * should be Nette\InvalidArgumentException, imho
* you need to call `!class_exist($templateClass) || ` before `!is_a` because `is_a` does not trigger autoloading

File: src/Bridges/ApplicationDI/LatteExtension.php [hunk 10]
```diff
@@ -52,9 +53,12 @@ public function loadConfiguration()
 			->addSetup('Nette\Utils\Html::$xhtml = ?', [(bool) $config['xhtml']])
 			->setImplement(Nette\Bridges\ApplicationLatte\ILatteFactory::class);
 
-		$builder->addDefinition($this->prefix('templateFactory'))
+		$templateFactory = $builder->addDefinition($this->prefix('templateFactory'))
 			->setClass(Nette\Application\UI\ITemplateFactory::class)
 			->setFactory(Nette\Bridges\ApplicationLatte\TemplateFactory::class);
+		if($config['templateClass'] !== NULL) {
```
Comment (JanTvrdik) [inline src/Bridges/ApplicationDI/LatteExtension.php]: missing space after if

**Comment by mrtnzlml** [comment]
Updated according to comments.

File: src/Bridges/ApplicationLatte/TemplateFactory.php [hunk 11]
```diff
@@ -112,4 +115,12 @@ public function createTemplate(UI\Control $control = NULL)
 		return $template;
 	}
 
+	public function setTemplateClass($templateClass)
+	{
+		if (!class_exists($templateClass) || !is_a($templateClass, UI\ITemplate::class, TRUE)) {
+			throw new \Nette\InvalidArgumentException('Template should be instance of ' . UI\ITemplate::class . ", $templateClass given.");
```
Comment (JanTvrdik) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: * `Nette\InvalidArgumentException`
* "Class '$templateClass' does not implement ... ."

File: src/Bridges/ApplicationLatte/TemplateFactory.php [hunk 12]
```diff
@@ -112,4 +115,12 @@ public function createTemplate(UI\Control $control = NULL)
 		return $template;
 	}
 
+	public function setTemplateClass($templateClass)
+	{
+		if (!class_exists($templateClass) || !is_a($templateClass, UI\ITemplate::class, TRUE)) {
```
Comment (dg) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: It should be instance of `Template::class`

Comment (mrtnzlml) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: Why not `ITemplate`? Can you please elaborate it more @dg? I don't fully understand... :) Thank you.

Comment (JanTvrdik) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: My guess is that it's because `TemplateFactory` has `@return Template` annotation. However Nette itself does not seem to rely on the instance being `Template`.

Comment (dg) [inline src/Bridges/ApplicationLatte/TemplateFactory.php]: It has `@return Template`, it assumes `__construct(Latte\Engine)` and `__set` for variables (or fields).

**Comment by mrtnzlml** [comment]
Rebased on top of master branch (because of failing tests).

**Comment by mrtnzlml** [comment]
Thank you @dg! :)
